File: src/db_savings.py
#!/usr/bin/env python3
"""Helper script to query token savings and tool call statistics as JSON."""
import sqlite3
import json
import sys
import os
import glob
import re
from datetime import datetime, timedelta

db_path = sys.argv[1] if len(sys.argv) > 1 else os.path.expanduser("~/.konoha/skills.db")

def parse_iso_datetime(dt_str):
    """Parse ISO datetime as a timezone-aware local-time value."""
    if not dt_str:
        return None
    dt_str = dt_str.strip().replace("Z", "+00:00")
    for fmt in ("%Y-%m-%dT%H:%M:%S.%f%z", "%Y-%m-%dT%H:%M:%S%z", "%Y-%m-%dT%H:%M:%S.%f", "%Y-%m-%dT%H:%M:%S"):
        try:
            dt = datetime.strptime(dt_str, fmt)
            if dt.tzinfo is None:
                dt = dt.astimezone()
            else:
                dt = dt.astimezone()
            return dt
        except ValueError:
            continue
    return None

def calculate_all_model_tokens():
    """Scan transcript files once to calculate generated content and thought tokens and USD costs for today, 7days, and alltime."""
    brain_dirs = [
        os.path.expanduser("~/.gemini/antigravity-cli/brain"),
        os.path.expanduser("~/.gemini/antigravity-ide/brain")
    ]
    patterns = [os.path.join(bd, "*", ".system_generated", "logs", "transcript.jsonl") for bd in brain_dirs]

    now = datetime.now()
    cutoff_today = now.replace(hour=0, minute=0, second=0, microsecond=0).astimezone()
    cutoff_7days = (now - timedelta(days=7)).replace(hour=0, minute=0, second=0, microsecond=0).astimezone()

    flash_out_rate = 0.30 / 1000000
    pro_out_rate = 5.00 / 1000000

    metrics = {
        "today": {"content_chars": 0, "thought_chars": 0, "cost": 0.0},
        "7days": {"content_chars": 0, "thought_chars": 0, "cost": 0.0},
        "all": {"content_chars": 0, "thought_chars": 0, "cost": 0.0},
    }

    all_paths = []
    for pattern in patterns:
        all_paths.extend(glob.glob(pattern, recursive=True))

    for path in all_paths:
        if not os.path.exists(path):
            continue
        try:
            with open(path, "r", encoding="utf-8", errors="ignore") as f:
                for line in f:
                    if not line.strip() or '"source":"MODEL"' not in line:
                        continue
                    try:
                        record = json.loads(line)
                        if record.get("source") == "MODEL" and record.get("type") == "PLANNER_RESPONSE":
                            created_at_str = record.get("created_at")
                            dt = parse_iso_datetime(created_at_str)

                            content = record.get("content") or ""
                            thinking = record.get("thinking") or ""
                            content_len = len(content)
                            thinking_len = len(thinking)
                            total_turn_out_tokens = (content_len + thinking_len) / 4.0

                            is_pro = True
                            lower_content = content.lower()
                            if "genin" in lower_content or "chunin" in lower_content or "tokubetsu" in lower_content or "jonin" in lower_content:
                                is_pro = False
                            elif "anbu" in lower_content or "kage" in lower_content or "antigravity" in lower_content:
                                is_pro = True
                            else:
                                if "genin" in path.lower() or "chunin" in path.lower() or "tokubetsu" in path.lower() or "jonin" in path.lower():
                                    is_pro = False

                            rate = pro_out_rate if is_pro else flash_out_rate
                            cost = total_turn_out_tokens * rate

                            metrics["all"]["content_chars"] += content_len
                            metrics["all"]["thought_chars"] += thinking_len
                            metrics["all"]["cost"] += cost

                            if dt and dt >= cutoff_7days:
                                metrics["7days"]["content_chars"] += content_len
                                metrics["7days"]["thought_chars"] += thinking_len
                                metrics["7days"]["cost"] += cost

                            if dt and dt >= cutoff_today:
                                metrics["today"]["content_chars"] += content_len
                                metrics["today"]["thought_chars"] += thinking_len
                                metrics["today"]["cost"] += cost
                    except Exception:
                        pass
        except Exception:
            pass

    return {
        key: {
            "content_tokens": int(metrics[key]["content_chars"] / 4),
            "thought_tokens": int(metrics[key]["thought_chars"] / 4),
            "output_cost_usd": metrics[key]["cost"]
        }
        for key in ("today", "7days", "all")
    }

def query_input_savings_cost(conn, time_filter=None):
    """Calculate input cost saved in USD based on agent model tiers."""
    where_clause = ""
    if time_filter == "today":
        where_clause = "WHERE date(timestamp, 'localtime') >= date('now', 'localtime')"
    elif time_filter == "7days":
        where_clause = "WHERE date(timestamp, 'localtime') >= date('now', '-7 days', 'localtime')"
        
    query = f"""
        SELECT agent, COALESCE(SUM(tokens_saved), 0) as tokens
        FROM tool_calls
        {where_clause}
        GROUP BY agent
    """
    rows = conn.execute(query).fetchall()
    
    flash_rate = 0.075 / 1000000
    pro_rate = 1.25 / 1000000
    total_saved_usd = 0.0
    
    for row in rows:
        agent = (row[0] or "").lower()
        tokens = row[1]
        
        is_pro = True
        if "genin" in agent or "chunin" in agent or "tokubetsu" in agent or "jonin" in agent:
            is_pro = False
            
        rate = pro_rate if is_pro else flash_rate
        total_saved_usd += tokens * rate
        
    return total_saved_usd

def query_stats(conn, time_filter=None, model_tokens=None):
    """Query statistics based on a SQL time filter."""
    where_clause = ""
    if time_filter is None or time_filter == "all":
        where_clause = ""
    elif time_filter == "today":
        where_clause = "WHERE date(timestamp, 'localtime') >= date('now', 'localtime')"
    elif time_filter == "7days":
        where_clause = "WHERE date(timestamp, 'localtime') >= date('now', '-7 days', 'localtime')"

    # Get actual library baseline size
    try:
        baseline_row = conn.execute("SELECT SUM(byte_size) FROM skills").fetchone()
        library_baseline_bytes = baseline_row[0] or 550000
    except Exception:
        library_baseline_bytes = 550000

    query = f"""
        SELECT
            COUNT(*) as calls,
            COALESCE(SUM(bytes_saved), 0) as bytes,
            COALESCE(SUM(tokens_saved), 0) as tokens,
            COALESCE(SUM(bytes_saved + returned_bytes), 0) as total_tool_calls_bytes
        FROM tool_calls
        {where_clause}
    """
    row = conn.execute(query).fetchone()
    total_bytes = row[3]
    pct = round((row[1] / total_bytes * 100)) if total_bytes > 0 else 0

    saved_usd = query_input_savings_cost(conn, time_filter)
    out_stats = model_tokens if model_tokens else {"content_tokens": 0, "thought_tokens": 0, "output_cost_usd": 0.0}
    net_saved_usd = max(0.0, saved_usd - out_stats["output_cost_usd"])
    
    # Query client breakdown
    by_client = {
        "antigravity": {"calls": 0, "bytes": 0, "tokens": 0},
        "agy": {"calls": 0, "bytes": 0, "tokens": 0},
        "cursor": {"calls": 0, "bytes": 0, "tokens": 0},
        "claudecode": {"calls": 0, "bytes": 0, "tokens": 0},
        "opencode": {"calls": 0, "bytes": 0, "tokens": 0},
        "commandcode": {"calls": 0, "bytes": 0, "tokens": 0},
        "codex": {"calls": 0, "bytes": 0, "tokens": 0},
    }
    
    query_client = f"""
        SELECT 
            COALESCE(client, 'antigravity') as c_name,
            COUNT(*) as calls,
            COALESCE(SUM(bytes_saved), 0) as bytes,
            COALESCE(SUM(tokens_saved), 0) as tokens
        FROM tool_calls
        {where_clause}
        GROUP BY c_name
    """
    try:
        rows_client = conn.execute(query_client).fetchall()
        for r_c in rows_client:
            c_name = r_c[0].lower()
            if c_name not in by_client:
                c_name = "unknown"
            by_client.setdefault(c_name, {"calls": 0, "bytes": 0, "tokens": 0})
            by_client[c_name]["calls"] += r_c[1]
            by_client[c_name]["bytes"] += r_c[2]
            by_client[c_name]["tokens"] += r_c[3]
    except Exception:
        pass
        
    return {
        "calls": row[0],
        "bytes": row[1],
        "tokens": row[2],
        "pct": pct,
        "total_bytes": total_bytes,
        "db_size_bytes": library_baseline_bytes,
        "saved_usd": saved_usd,
        "content_tokens": out_stats["content_tokens"],
        "thought_tokens": out_stats["thought_tokens"],
        "output_cost_usd": out_stats["output_cost_usd"],
        "net_saved_usd": net_saved_usd,
        "by_client": by_client
    }

def sanitize_legacy_records(conn):
    """Sanitize legacy inflated baseline records and remove test records."""
    try:
        conn.execute("DELETE FROM tool_calls WHERE tool LIKE 'test_%' OR query LIKE '%test_tool%'")
        conn.execute("""
            UPDATE tool_calls
            SET total_library_bytes = returned_bytes,
                bytes_saved = 0,
                tokens_saved = 0
            WHERE tool NOT IN ('find_skill', 'list_skills')
              AND total_library_bytes >= 400000
        """)
        conn.commit()
    except Exception:
        pass

try:
    if not os.path.exists(db_path):
        print(json.dumps({"error": f"Database not found at {db_path}"}))
        sys.exit(1)
    
    conn = sqlite3.connect(db_path)

    # Ensure table exists (agent/client included in CREATE to avoid redundant ALTER)
    conn.execute("""
        CREATE TABLE IF NOT EXISTS tool_calls (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            timestamp DATETIME DEFAULT CURRENT_TIMESTAMP,
            tool TEXT NOT NULL,
            query TEXT,
            returned_bytes INTEGER,
            total_library_bytes INTEGER,
            bytes_saved INTEGER,
            tokens_saved INTEGER,
            agent TEXT,
            client TEXT
        );
    """)

    sanitize_legacy_records(conn)

    all_model_tokens = calculate_all_model_tokens()
    stats_today = query_stats(conn, "today", all_model_tokens.get("today"))
    stats_7days = query_stats(conn, "7days", all_model_tokens.get("7days"))
    stats_all = query_stats(conn, "all", all_model_tokens.get("all"))
    
    # Query tool call breakdown by type
    def query_by_call_type(c):
        q = """
            SELECT 
                tool,
                COUNT(*) as calls,
                COALESCE(SUM(bytes_saved), 0) as bytes,
                COALESCE(SUM(bytes_saved + returned_bytes), 0) as total_bytes
            FROM tool_calls
            GROUP BY tool
            ORDER BY calls DESC
        """
        rows = c.execute(q).fetchall()
        results = []
        for r in rows:
            tool = r[0]
            calls = r[1]
            bytes_saved = r[2]
            total_bytes = r[3]
            pct = round((bytes_saved / total_bytes * 100)) if total_bytes > 0 else 0
            results.append({
                "tool": tool,
                "calls": calls,
                "bytes": bytes_saved,
                "pct": pct
            })
        return results

    by_call_type = query_by_call_type(conn)
    conn.close()
    
    print(json.dumps({
        "today": stats_today,
        "last7days": stats_7days,
        "alltime": stats_all,
        "by_call_type": by_call_type
    }))
except Exception as e:
    print(json.dumps({"error": str(e)}))
    sys.exit(1)

File: src/db_savings.js
#!/usr/bin/env node
/**
 * Helper script and in-process module to query token savings and tool call statistics as JSON.
 * Pure Node.js replacement for db_savings.py using better-sqlite3.
 */

const fs = require('fs');
const path = require('path');
const os = require('os');
const db = require('./db');

function parseIsoDatetime(dtStr) {
  if (!dtStr) return null;
  const d = new Date(dtStr);
  return isNaN(d.getTime()) ? null : d;
}

function calculateAllModelTokens() {
  const home = os.homedir();
  const brainDirs = [
    path.join(home, '.gemini', 'antigravity-cli', 'brain'),
    path.join(home, '.gemini', 'antigravity-ide', 'brain')
  ];

  const now = new Date();
  const cutoffToday = new Date(now.getFullYear(), now.getMonth(), now.getDate(), 0, 0, 0, 0);
  const cutoff7Days = new Date(now.getFullYear(), now.getMonth(), now.getDate() - 7, 0, 0, 0, 0);
  const cutoff7DaysTs = cutoff7Days.getTime() / 1000;

  const flashOutRate = 0.30 / 1000000;
  const proOutRate = 5.00 / 1000000;

  const metrics = {
    today: { content_chars: 0, thought_chars: 0, cost: 0.0 },
    "7days": { content_chars: 0, thought_chars: 0, cost: 0.0 },
    all: { content_chars: 0, thought_chars: 0, cost: 0.0 }
  };

  const cacheFile = path.join(home, '.konoha', 'transcript_cache.json');
  let cache = { version: 1, files: {} };
  if (fs.existsSync(cacheFile)) {
    try {
      cache = JSON.parse(fs.readFileSync(cacheFile, 'utf-8'));
      if (!cache || typeof cache !== 'object' || !cache.files) {
        cache = { version: 1, files: {} };
      }
    } catch (_) {
      cache = { version: 1, files: {} };
    }
  }

  // Find all transcript.jsonl files
  const allPaths = [];
  for (const bDir of brainDirs) {
    if (!fs.existsSync(bDir)) continue;
    try {
      const convEntries = fs.readdirSync(bDir);
      for (const conv of convEntries) {
        const tPath = path.join(bDir, conv, '.system_generated', 'logs', 'transcript.jsonl');
        if (fs.existsSync(tPath)) {
          allPaths.push(tPath);
        }
      }
    } catch (_) { /* intentional best-effort fallback: failure here must never crash the CLI/MCP runtime */ }
  }

  let cacheUpdated = false;
  for (const filePath of allPaths) {
    if (!fs.existsSync(filePath)) continue;
    try {
      const st = fs.statSync(filePath);
      const mtime = st.mtimeMs / 1000;
      const size = st.size;

      if (mtime < cutoff7DaysTs) {
        const cachedEntry = cache.files && cache.files[filePath];
        if (cachedEntry && Math.abs(cachedEntry.mtime - mtime) < 0.001 && cachedEntry.size === size) {
          const cAll = cachedEntry.all || {};
          metrics.all.content_chars += (cAll.content_chars || 0);
          metrics.all.thought_chars += (cAll.thought_chars || 0);
          metrics.all.cost += (cAll.cost || 0.0);
          continue;
        }
      }

      const fileAll = { content_chars: 0, thought_chars: 0, cost: 0.0 };
      const contentStr = fs.readFileSync(filePath, 'utf-8');
      const lines = contentStr.split('\n');

      for (const line of lines) {
        const trimmed = line.trim();
        if (!trimmed || !trimmed.includes('"source":"MODEL"')) continue;
        try {
          const record = JSON.parse(trimmed);
          if (record.source === "MODEL" && record.type === "PLANNER_RESPONSE") {
            const dt = parseIsoDatetime(record.created_at);
            const content = record.content || "";
            const thinking = record.thinking || "";
            const contentLen = content.length;
            const thinkingLen = thinking.length;
            const totalTurnOutTokens = (contentLen + thinkingLen) / 4.0;

            let isPro = true;
            const lowerContent = content.toLowerCase();
            if (lowerContent.includes("genin") || lowerContent.includes("chunin") || lowerContent.includes("tokubetsu") || lowerContent.includes("jonin")) {
              isPro = false;
            } else if (lowerContent.includes("anbu") || lowerContent.includes("kage") || lowerContent.includes("antigravity")) {
              isPro = true;
            } else {
              const lowerPath = filePath.toLowerCase();
              if (lowerPath.includes("genin") || lowerPath.includes("chunin") || lowerPath.includes("tokubetsu") || lowerPath.includes("jonin")) {
                isPro = false;
              }
            }

            const rate = isPro ? proOutRate : flashOutRate;
            const cost = totalTurnOutTokens * rate;

            fileAll.content_chars += contentLen;
            fileAll.thought_chars += thinkingLen;
            fileAll.cost += cost;

            metrics.all.content_chars += contentLen;
            metrics.all.thought_chars += thinkingLen;
            metrics.all.cost += cost;

            if (dt && dt >= cutoff7Days) {
              metrics["7days"].content_chars += contentLen;
              metrics["7days"].thought_chars += thinkingLen;
              metrics["7days"].cost += cost;
            }

            if (dt && dt >= cutoffToday) {
              metrics.today.content_chars += contentLen;
              metrics.today.thought_chars += thinkingLen;
              metrics.today.cost += cost;
            }
          }
        } catch (_) { /* intentional best-effort fallback: failure here must never crash the CLI/MCP runtime */ }
      }

      if (!cache.files) cache.files = {};
      cache.files[filePath] = { mtime, size, all: fileAll };
      cacheUpdated = true;
    } catch (_) { /* intentional best-effort fallback: failure here must never crash the CLI/MCP runtime */ }
  }

  if (cacheUpdated) {
    try {
      const dir = path.dirname(cacheFile);
      if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
      fs.writeFileSync(cacheFile, JSON.stringify(cache), 'utf-8');
    } catch (_) { /* intentional best-effort fallback: failure here must never crash the CLI/MCP runtime */ }
  }

  return {
    today: {
      content_tokens: Math.floor(metrics.today.content_chars / 4),
      thought_tokens: Math.floor(metrics.today.thought_chars / 4),
      output_cost_usd: metrics.today.cost
    },
    "7days": {
      content_tokens: Math.floor(metrics["7days"].content_chars / 4),
      thought_tokens: Math.floor(metrics["7days"].thought_chars / 4),
      output_cost_usd: metrics["7days"].cost
    },
    all: {
      content_tokens: Math.floor(metrics.all.content_chars / 4),
      thought_tokens: Math.floor(metrics.all.thought_chars / 4),
      output_cost_usd: metrics.all.cost
    }
  };
}

function queryInputSavingsCost(conn, timeFilter = null) {
  let whereClause = "";
  if (timeFilter === "today") {
    whereClause = "WHERE date(timestamp, 'localtime') >= date('now', 'localtime')";
  } else if (timeFilter === "7days") {
    whereClause = "WHERE date(timestamp, 'localtime') >= date('now', '-7 days', 'localtime')";
  }

  const query = `
    SELECT agent, COALESCE(SUM(tokens_saved), 0) as tokens
    FROM tool_calls
    ${whereClause}
    GROUP BY agent
  `;
  const rows = conn.prepare(query).all();

  const flashRate = 0.075 / 1000000;
  const proRate = 1.25 / 1000000;
  let totalSavedUsd = 0.0;

  for (const row of rows) {
    const agent = (row.agent || "").toLowerCase();
    const tokens = row.tokens || 0;

    let isPro = true;
    if (agent.includes("genin") || agent.includes("chunin") || agent.includes("tokubetsu") || agent.includes("jonin")) {
      isPro = false;
    }

    const rate = isPro ? proRate : flashRate;
    totalSavedUsd += tokens * rate;
  }

  return totalSavedUsd;
}

function queryStats(conn, timeFilter = null, modelTokens = null) {
  let whereClause = "";
  if (timeFilter === null || timeFilter === "all") {
    whereClause = "";
  } else if (timeFilter === "today") {
    whereClause = "WHERE date(timestamp, 'localtime') >= date('now', 'localtime')";
  } else if (timeFilter === "7days") {
    whereClause = "WHERE date(timestamp, 'localtime') >= date('now', '-7 days', 'localtime')";
  }

  let libraryBaselineBytes = 2065 * 1024;
  try {
    const baselineRow = conn.prepare("SELECT SUM(byte_size) as b FROM skills").get();
    if (baselineRow && baselineRow.b) {
      libraryBaselineBytes = 2065 * 1024;
    }
  } catch (_) { /* intentional best-effort fallback: failure here must never crash the CLI/MCP runtime */ }

  const query = `
    SELECT
        COUNT(*) as calls,
        COALESCE(SUM(bytes_saved), 0) as bytes,
        COALESCE(SUM(tokens_saved), 0) as tokens,
        COALESCE(SUM(bytes_saved + returned_bytes), 0) as total_tool_calls_bytes
    FROM tool_calls
    ${whereClause}
  `;
  const row = conn.prepare(query).get();
  const totalBytes = row.total_tool_calls_bytes;
  const pct = totalBytes > 0 ? Math.round((row.bytes / totalBytes) * 100) : 0;

  const savedUsd = queryInputSavingsCost(conn, timeFilter);
  const outStats = modelTokens || { content_tokens: 0, thought_tokens: 0, output_cost_usd: 0.0 };
  const netSavedUsd = Math.max(0.0, savedUsd - outStats.output_cost_usd);

  const byClient = {
    antigravity: { calls: 0, bytes: 0, tokens: 0 },
    agy: { calls: 0, bytes: 0, tokens: 0 },
    cursor: { calls: 0, bytes: 0, tokens: 0 },
    claudecode: { calls: 0, bytes: 0, tokens: 0 },
    opencode: { calls: 0, bytes: 0, tokens: 0 },
    commandcode: { calls: 0, bytes: 0, tokens: 0 },
    codex: { calls: 0, bytes: 0, tokens: 0 },
    pi: { calls: 0, bytes: 0, tokens: 0 },
    // Calls with no verified client session signal (honest attribution):
    // displayed as "Unattributed" instead of being hidden or labeled "Unknown".
    unattributed: { calls: 0, bytes: 0, tokens: 0 },
  };

  const queryClient = `
    SELECT
        COALESCE(client, 'antigravity') as c_name,
        COUNT(*) as calls,
        COALESCE(SUM(bytes_saved), 0) as bytes,
        COALESCE(SUM(tokens_saved), 0) as tokens
    FROM tool_calls
    ${whereClause}
    GROUP BY c_name
  `;

  try {
    const rowsClient = conn.prepare(queryClient).all();
    for (const rC of rowsClient) {
      let cName = (rC.c_name || "").toLowerCase();
      if (!(cName in byClient)) {
        cName = "unattributed";
      }
      if (!byClient[cName]) {
        byClient[cName] = { calls: 0, bytes: 0, tokens: 0 };
      }
      byClient[cName].calls += rC.calls;
      byClient[cName].bytes += rC.bytes;
      byClient[cName].tokens += rC.tokens;
    }
  } catch (_) { /* intentional best-effort fallback: failure here must never crash the CLI/MCP runtime */ }

  return {
    calls: row.calls,
    bytes: row.bytes,
    tokens: row.tokens,
    pct,
    total_bytes: totalBytes,
    db_size_bytes: libraryBaselineBytes,
    saved_usd: savedUsd,
    content_tokens: outStats.content_tokens,
    thought_tokens: outStats.thought_tokens,
    output_cost_usd: outStats.output_cost_usd,
    net_saved_usd: netSavedUsd,
    by_client: byClient
  };
}

function sanitizeLegacyRecords(conn) {
  try {
    conn.prepare("DELETE FROM tool_calls WHERE tool LIKE 'test_%' OR query LIKE '%test_tool%'").run();
    conn.prepare(`
      UPDATE tool_calls
      SET total_library_bytes = returned_bytes,
          bytes_saved = 0,
          tokens_saved = 0
      WHERE tool IN ('anbu', 'genin', 'sannin', 'kage', 'jonin', 'tokubetsu_jonin', 'tokubetsu-jonin')
        AND total_library_bytes > returned_bytes
    `).run();
    conn.prepare(`
      UPDATE tool_calls
      SET total_library_bytes = returned_bytes,
          bytes_saved = 0,
          tokens_saved = 0
      WHERE tool NOT IN ('find_skill', 'find_skills', 'list_skills', 'optimize_report', 'get_skill')
        AND total_library_bytes >= 400000
    `).run();
    conn.prepare(`
      UPDATE tool_calls
      SET total_library_bytes = MAX(returned_bytes, 15000),
          bytes_saved = MAX(0, MAX(returned_bytes, 15000) - returned_bytes),
          tokens_saved = CAST((MAX(0, MAX(returned_bytes, 15000) - returned_bytes) / 4) AS INTEGER)
      WHERE tool IN ('find_skill', 'find_skills')
        AND total_library_bytes >= 400000
    `).run();
    // Normalize legacy list_skills / optimize_report rows from old full-library 2.11 MB baseline down to realistic 35,000-byte frontmatters baseline
    conn.prepare(`
      UPDATE tool_calls
      SET total_library_bytes = MAX(returned_bytes, 35000),
          bytes_saved = MAX(0, MAX(returned_bytes, 35000) - returned_bytes),
          tokens_saved = CAST((MAX(0, MAX(returned_bytes, 35000) - returned_bytes) / 4) AS INTEGER)
      WHERE tool IN ('list_skills', 'optimize_report')
        AND total_library_bytes >= 400000
    `).run();
    try {
      const getSkillRows = conn.prepare("SELECT id, query, returned_bytes FROM tool_calls WHERE tool = 'get_skill' AND bytes_saved = 0").all();
      const updateGetSkill = conn.prepare("UPDATE tool_calls SET total_library_bytes = ?, bytes_saved = ?, tokens_saved = ? WHERE id = ?");
      for (const gRow of getSkillRows) {
        let skillTarget = gRow.query;
        if (typeof skillTarget === 'string' && skillTarget.startsWith('{')) {
          try { skillTarget = JSON.parse(skillTarget).name || skillTarget; } catch (_) { /* best-effort json parse fallback */ }
        }
        const sRow = conn.prepare('SELECT byte_size FROM skills WHERE name = ? OR name LIKE ? LIMIT 1').get(skillTarget, `%/${skillTarget}`);
        if (sRow && sRow.byte_size > gRow.returned_bytes) {
          const bSaved = sRow.byte_size - gRow.returned_bytes;
          const tSaved = Math.floor(bSaved / 4);
          updateGetSkill.run(sRow.byte_size, bSaved, tSaved, gRow.id);
        }
      }
    } catch (_) { /* ignore */ }
  } catch (_) { /* intentional best-effort fallback: failure here must never crash the CLI/MCP runtime */ }
}

function getSavingsReport(dbPath = null) {
  const targetPath = dbPath || db.DB_PATH;
  if (!fs.existsSync(targetPath)) {
    throw new Error(`Database not found at ${targetPath}`);
  }

  const conn = db.getConnection(targetPath, false);
  try {
    db.setupSchema(conn);
    sanitizeLegacyRecords(conn);

    const allModelTokens = calculateAllModelTokens();
    const statsToday = queryStats(conn, "today", allModelTokens.today);
    const stats7Days = queryStats(conn, "7days", allModelTokens["7days"]);
    const statsAll = queryStats(conn, "all", allModelTokens.all);

    const rows = conn.prepare(`
      SELECT
          tool,
          COUNT(*) as calls,
          COALESCE(SUM(bytes_saved), 0) as bytes,
          COALESCE(SUM(bytes_saved + returned_bytes), 0) as total_bytes
      FROM tool_calls
      GROUP BY tool
      ORDER BY calls DESC
    `).all();

    const byCallType = [];
    for (const r of rows) {
      const calls = r.calls;
      const bytesSaved = r.bytes;
      const totalBytes = r.total_bytes;
      const pct = totalBytes > 0 ? Math.round((bytesSaved / totalBytes) * 100) : 0;
      byCallType.push({
        tool: r.tool,
        calls,
        bytes: bytesSaved,
        pct
      });
    }

    let tiktokenAccuracy = null;
    try {
      const tokenSampler = require('./token_sampler');
      tiktokenAccuracy = tokenSampler.getDriftMetrics();
    } catch (_) { /* best-effort fallback: failure here must never crash telemetry */ }

    return {
      today: statsToday,
      last7days: stats7Days,
      alltime: statsAll,
      by_call_type: byCallType,
      tiktoken_accuracy: tiktokenAccuracy
    };
  } finally {
    conn.close();
  }
}

if (require.main === module) {
  const customDbPath = process.argv[2] || null;
  try {
    const report = getSavingsReport(customDbPath);
    console.log(JSON.stringify(report));
  } catch (err) {
    console.log(JSON.stringify({ error: err.message }));
    process.exit(1);
  }
}

module.exports = {
  getSavingsReport,
  calculateAllModelTokens,
  queryStats,
  queryInputSavingsCost,
  sanitizeLegacyRecords
};
